### PR TITLE
[thanos] support objstore.config=type:s3 for v0.2.0

### PR DESCRIPTION
**Pull request by gangseok514**
from v0.2.0, thanos removed --s3.xxx options and added objstore

thanos value in prometheus crd changes to objstore values in thanos-sidecar

Signed-off-by: gangseok.lee <gangseok.lee@samsung.com>

### CODE REVIEW

**Comment by gangseok514** [comment]
@metalmatze Hello, This is my first PR. would you review this? Thanks!

**Comment by gangseok514** [comment]
@metalmatze Hello, I'd like to hear some little advice. :)

**Comment by abursavich** [comment]
I would suggest leaving the existing code as is, marking the old fields as deprecated, and adding a new field to the ThanosSpec, e.g.:

```go
type ThanosSpec struct {
    // ... other fields omitted ...
    // Deprecated: Use ObjectStorageConfig for thanos v0.2.0+
    GCS *ThanosGCSSpec `json:"gcs,omitempty"`
    // Deprecated: Use ObjectStorageConfig for thanos v0.2.0+
    S3 *ThanosS3Spec `json:"s3,omitempty"`
    // Object storage configuration.
    ObjectStorageConfig *v1.SecretKeySelector `json:"objectStorageConfig,omitempty"`
}
```

```go
if p.Spec.Thanos.ObjectStorageConfig != nil {
    envVars = append(envVars, v1.EnvVar{
        Name: "OBJECT_STORAGE_CONFIG",
        ValueFrom: &v1.EnvVarSource{
            SecretKeyRef: p.Spec.Thanos.ObjectStorageConfig,
        },
    })
    thanosArgs = append(thanosArgs, "--objstore.config", "$(OBJECT_STORAGE_CONFIG)")
}
```

The main benefit is that prometheus-operator doesn't have to explicitly support every thanos storage backend (GCS, S3, Azure, Swift, etc.).

It also allows users to manage their own upgrade path. It would be possible for someone to pick up a new version of prometheus-operator and stick with an older version of thanos until they're ready. 

Eventually, the `GCS` and `S3` fields could be removed from `ThanosSpec` an older versions of thanos can be abandoned.

**Comment by abursavich** [comment]
Actually, I found an existing issue for Azure storage that supports something like what I suggested: https://github.com/coreos/prometheus-operator/issues/2133#issuecomment-438989379

**Comment by mmack** [comment]
This works for me.
Dockerimage: https://cloud.docker.com/u/mmack/repository/docker/mmack/prometheus-operator
Commit: https://github.com/mackIOConsulting/prometheus-operator/commit/10481d5c5622149beee3275428882438d72659a6

File: pkg/prometheus/statefulset.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ import (
 const (
 	governingServiceName     = "prometheus-operated"
 	DefaultPrometheusVersion = "v2.5.0"
-	DefaultThanosVersion     = "v0.1.0"
+	DefaultThanosVersion     = "v0.2.0"
```
Comment (bwplotka) [inline pkg/prometheus/statefulset.go]: Can we use 0.2.1?

File: pkg/prometheus/statefulset.go [hunk 2]
```diff
@@ -699,9 +705,7 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *Config, ruleConfigMapName
 						SecretKeyRef: p.Spec.Thanos.S3.AccessKey,
 					},
 				})
-			}
-			if p.Spec.Thanos.S3.EncryptSSE != nil && *p.Spec.Thanos.S3.EncryptSSE {
-				thanosArgs = append(thanosArgs, "--s3.encrypt-sse")
+				objstore = append(objstore, "  access_key: $(S3_ACCESS_KEY)")
```
Comment (bwplotka) [inline pkg/prometheus/statefulset.go]: I guess those are defined? They must be in pod spec env vars for container.

**Comment by metalmatze** [comment]
I'll create a unit test for the generation.
I was thinking about generating a secret with the config would be helpful, as that secret could also be used by the store component for example.

**Comment by bwplotka** [comment]
Up to you, we use compacted  format as it is just nicer to read all configuration from flags. (:

**Comment by abursavich** [comment]
Please let users create their own secret with the configuration that they reference in the ThanoSpec. Users can reference the same secret for the store and compactor, but most importantly, it decouples the thanos storage backend config from the prometheus-operator config. Thanos will be able to support arbitrary backends without needed to push changes to prometheus-operator.

File: pkg/prometheus/statefulset.go (hunk 2 above)
Comment (gangseok514) [reply pkg/prometheus/statefulset.go]: the above of it...

```
                          if p.Spec.Thanos.S3.AccessKey != nil {
				envVars = append(envVars, v1.EnvVar{
					Name: "S3_ACCESS_KEY",
					ValueFrom: &v1.EnvVarSource{
						SecretKeyRef: p.Spec.Thanos.S3.AccessKey,
					},
				})
			}
```

is defiend.

File: pkg/prometheus/statefulset.go (hunk 1 above)
Comment (gangseok514) [reply pkg/prometheus/statefulset.go]: Done!

**Comment by gangseok514** [comment]
I think @abursavich's approach(https://github.com/coreos/prometheus-operator/pull/2264) is better than my changes. But the approach contains "not secret" data. containing only secret data in Secret resource is more proper use.

**Comment by gangseok514** [comment]
@metalmatze I think that separating the secret from prometheus-operator is better. because as @abursavich  said that the secret will be shared by thanos-store/compactor.

**Comment by mmack** [comment]
Yeah this is the better approach!

**Comment by brancz** [comment]
cc @metalmatze

**Comment by metalmatze** [comment]
Perfect. Referecing a secret with the config seems like a good idea, agree. 
Let's continue with #2264 and discuss there.

Thanks for the initiative @gangseok514 :+1:

**Pull request closed without merging**